File: backend/src/main/scala/slaydemo/backend/battle/objects/BattleWeaponThermalState.scala
package slaydemo.backend.battle.objects

enum BattleWeaponThermalState {
  case Ready
  case Overheated(remainingMs: CooldownMillis)
}

object BattleWeaponThermalState {
  def overheatedFlag(value: BattleWeaponThermalState): Boolean =
    value match {
      case BattleWeaponThermalState.Overheated(_) => true
      case BattleWeaponThermalState.Ready         => false
    }

  def overheatRemainingMs(value: BattleWeaponThermalState): CooldownMillis =
    value match {
      case BattleWeaponThermalState.Ready =>
        CooldownMillis(0)
      case BattleWeaponThermalState.Overheated(remainingMs) =>
        CooldownMillis(math.max(0, remainingMs.value))
    }

  def overheated(remainingMs: CooldownMillis): BattleWeaponThermalState =
    if remainingMs.value <= 0 then BattleWeaponThermalState.Ready
    else BattleWeaponThermalState.Overheated(CooldownMillis(remainingMs.value))
}

File: frontend/src/objects/battle/microservices/combat/objects/weapon/BattleWeaponThermalState.ts
export type BattleWeaponThermalState =
  | { status: "ready" }
  | { status: "overheated"; remainingMs: number };

export function overheatedFlag(value: BattleWeaponThermalState): boolean {
  return value.status === "overheated";
}

export function overheatRemainingMs(value: BattleWeaponThermalState): number {
  return value.status === "overheated" ? Math.max(0, Math.round(value.remainingMs)) : 0;
}

export function overheated(remainingMs: number): BattleWeaponThermalState {
  return remainingMs <= 0 ? { status: "ready" } : { status: "overheated", remainingMs: Math.round(remainingMs) };
}

